dbms/src/DataStreams/MultiplexInputStream.h

```diff
@@ -0,0 +1,259 @@
+        if (queue_id != static_cast<int>(streams_queue_by_partition.size()) - 1)
+        {
+            swap(streams_queue_by_partition[queue_id], streams_queue_by_partition.back());
+            streams_queue_by_partition.pop_back();
+            return queue_id;
+        }
+        else
+        {
+            streams_queue_by_partition.pop_back();
+            return 0;
+        }
+    }
+
+    int nextQueueId(int queue_id) const
+    {
+        if (queue_id + 1 < static_cast<int>(streams_queue_by_partition.size()))
+            return queue_id + 1;
+        else
+            return 0;
+    }
+
+    static void swap(std::shared_ptr<std::queue<std::shared_ptr<IBlockInputStream>>> & a,
+                     std::shared_ptr<std::queue<std::shared_ptr<IBlockInputStream>>> & b)
+    {
+        auto tmp = a;
+        a = b;
+        b = tmp;
+    }
+
+    std::vector<
+        std::shared_ptr<std::queue<
+            std::shared_ptr<IBlockInputStream>>>>
+        streams_queue_by_partition;
+    std::vector<std::shared_ptr<IBlockInputStream>> added_streams;
+    int streams_queue_id = 0;
+    std::mutex mu;
+};
+
+class MultiplexInputStream final : public IProfilingBlockInputStream
+{
+private:
+    static constexpr auto NAME = "Multiplex";
+
+public:
+    MultiplexInputStream(
+        std::shared_ptr<MultiPartitionStreamPool> & shared_pool,
+        const String & req_id)
+        : log(Logger::get(NAME, req_id))
+        , shared_pool(shared_pool)
+    {
+        shared_pool->exportAddedStreams(children);
+        size_t num_children = children.size();
+        if (num_children > 1)
+        {
+            Block header = children.at(0)->getHeader();
+            for (size_t i = 1; i < num_children; ++i)
+                assertBlocksHaveEqualStructure(
+                    children[i]->getHeader(),
+                    header,
+                    "MULTIPLEX");
+        }
+    }
+
+    String getName() const override { return NAME; }
+
+    ~MultiplexInputStream() override
+    {
+        try
+        {
+            if (!all_read)
+                cancel(false);
+        }
+        catch (...)
+        {
+            tryLogCurrentException(log, __PRETTY_FUNCTION__);
+        }
+    }
+
+    /** Different from the default implementation by trying to stop all sources,
+      * skipping failed by execution.
+      */
+    void cancel(bool kill) override
+    {
+        if (kill)
+            is_killed = true;
+
+        bool old_val = false;
+        if (!is_cancelled.compare_exchange_strong(
+                old_val,
+                true,
+                std::memory_order_seq_cst,
+                std::memory_order_relaxed))
+            return;
+
+        if (cur_stream)
+        {
+            if (IProfilingBlockInputStream * child = dynamic_cast<IProfilingBlockInputStream *>(&*cur_stream))
+            {
+                child->cancel(kill);
+            }
+        }
+    }
+
+    BlockExtraInfo getBlockExtraInfo() const override
```

Comment:
Why override instead of using default implementation?
```
virtual BlockExtraInfo getBlockExtraInfo() const
{
        throw Exception("Method getBlockExtraInfo is not supported by the data stream " + getName(), ErrorCodes::NOT_IMPLEMENTED);
}
```

Reply:
you can refer to the UnionBlockInputStream

Reply:
good idea